added solution

JohnGrey17 opened the pull request:
Implemented Validator for DTO and CustomGlobalExceptionHandler with implementation for validator and EntityNotFoundException

JohnGrey17 committed:
removed BookRepository implementation Class instead that did extend JpaRepositoty Interface

JohnGrey17 committed:
changer parameter im spring.jpa.hibernate.ddl-auto on -validate

JohnGrey17 committed:
added changeLog configuration

JohnGrey17 committed:
Implemented PUT request

JohnGrey17 committed:
implemented soft delete (added isdeleted to Entity and yaml file)
implemented method on service layer and controller

JohnGrey17 committed:
implemented soft delete (added isdeleted to Entity and yaml file)
implemented method on service layer and controller

JohnGrey17 committed:
implemented soft delete (added isdeleted to Entity and yaml file)
implemented method on service layer and controller

JohnGrey17 committed:
fixed update method in service layer
changed deprecated @WHERE annotation on @SQLRestriction

JohnGrey17 committed:
MVN clean package

JohnGrey17 committed:
added environment variable dor password and login

JohnGrey17 committed:
fixed implementation for updateBookById in service layer

JohnGrey17 committed:
fixed solution

JohnGrey17 committed:
Added BookSearchParameter class to DTO folder
added specificationBuilderInterface
added bookSpecificationBuilderImpl(without implementation)

JohnGrey17 committed:
Added BookSearchParameter class to DTO folder
added specificationBuilderInterface
added bookSpecificationBuilderImpl(without implementation)

JohnGrey17 committed:
implemented search method

JohnGrey17 committed:
implemented search method

JohnGrey17 committed:
fixed search method

JohnGrey17 committed:
Implemented Validator for DTO and CustomGlobalExceptionHandler with implementation for validator and EntityNotFoundException

TonyH277 commented on src/main/java/org/example/bookstore/dto/BookDto.java:
```diff
@@ -1,14 +1,21 @@
 @Data
 public class BookDto {
     private long id;
+    @NotNull
     private String title;
+    @NotNull
     private String author;
+    @NotNull
     private String isbn;
+    @NotNull
+    @Min(0)
```
do you need constraints here?

TonyH277 commented on src/main/resources/db/changelog/db.changelog-master.yaml:
```diff
@@ -1,3 +1,5 @@
   - include:
       file: db/changelog/changes/01-create-books-table.yaml
+
+
```
```suggestion
```

TonyH277 reviewed (approved):
LGFM! Left minor comments

JohnGrey17 committed:
Fixed solution for Validator

andrii-hoienko commented on pom.xml:
```diff
@@ -70,6 +70,15 @@
             <artifactId>liquibase-maven-plugin</artifactId>
             <version>${liquibase.version}</version>
         </dependency>
+        <dependency>
+            <groupId>org.hibernate.validator</groupId>
+            <artifactId>hibernate-validator</artifactId>
+            <version>7.0.1.Final</version>
+        </dependency>
```
Why this dependency is required?

andrii-hoienko commented on src/main/java/org/example/bookstore/controller/BookControllerImpl.java:
```diff
@@ -36,7 +35,8 @@ public BookDto getBookById(@PathVariable Long id) {
     @PostMapping
-    public BookDto createBook(@RequestBody BookRequestDto requestDto) {
+    @ResponseStatus(HttpStatus.CREATED)
+    public BookDto createBook(@RequestBody @Valid BookRequestDto requestDto) {
```
I believe you should validate on update as well

andrii-hoienko commented on src/main/java/org/example/bookstore/dto/BookRequestDto.java:
```diff
@@ -1,13 +1,22 @@
     private String title;
+    @NotNull
     private String author;
+    @NotNull
```
```suggestion
    @ISBN
```

JohnGrey17 replied:
> 

please clarify what i need to change here

karma-o replied:
better to ask those questions on the platform or QnA next time, because we rarely check on you questions here before the actual review

andrii-hoienko commented on src/main/java/org/example/bookstore/exception/CustomGlobalExceptionHandler.java:
```diff
@@ -0,0 +1,57 @@
+        return new ResponseEntity<>(body, HttpStatus.NOT_FOUND);
+    }
+
+    @Override
```
Override methods should go first

andrii-hoienko commented on src/main/java/org/example/bookstore/exception/CustomGlobalExceptionHandler.java:
```diff
@@ -0,0 +1,57 @@
+    @ExceptionHandler(EntityNotFoundException.class)
+    protected ResponseEntity<Object> handleEntityNotFoundException(EntityNotFoundException ex) {
+        Map<String, Object> body = new LinkedHashMap<>();
+        body.put("timestamp", LocalDate.now());
+        body.put("status", HttpStatus.NOT_FOUND);
+        body.put("error", "Entity not found");
+        body.put("message", ex.getMessage());
```
Code duplication, I suggest creating method, which will accept status code and error message and will return error message body

andrii-hoienko commented on src/main/java/org/example/bookstore/exception/CustomGlobalExceptionHandler.java:
```diff
@@ -0,0 +1,57 @@
+@ControllerAdvice
+public class CustomGlobalExceptionHandler extends ResponseEntityExceptionHandler {
+
+    @ExceptionHandler(EntityNotFoundException.class)
```
Only EntityNotFoundException could occur in your code?

andrii-hoienko commented on src/main/resources/application.properties:
```diff
@@ -8,3 +8,4 @@ spring.jpa.hibernate.ddl-auto=validate
 spring.jpa.show-sql=true
 spring.jpa.open-in-view=false
 spring.jackson.deserialization.fail-on-unknown-properties=true
+
```
```suggestion
```

andrii-hoienko reviewed (changes requested):
Good job, but some improvements required

JohnGrey17 committed:
Fixed solution

karma-o commented on src/main/java/org/example/bookstore/dto/BookRequestDto.java:
```diff
@@ -1,13 +1,21 @@
+    @NotBlank
     private String isbn;
+    @DecimalMin(value = "0.0", inclusive = false)
+    @Digits(integer = 3, fraction = 2)
```
I don't think we want to limit the price to the max of `999.99`. What if we would want to sell some rare books that could cost way more than that?

karma-o commented on src/main/java/org/example/bookstore/exception/InvalidPriceException.java:
```diff
@@ -0,0 +1,7 @@
+package org.example.bookstore.exception;
+
+public class InvalidPriceException extends RuntimeException {
```
I do not see any usage of that exception except in the `ExceptionHandler`, so let's remove that

karma-o commented on src/main/java/org/example/bookstore/exception/CustomGlobalExceptionHandler.java:
```diff
@@ -0,0 +1,71 @@
+        return e.getDefaultMessage();
+    }
+
+    @ExceptionHandler(EntityNotFoundException.class)
+    protected ResponseEntity<Object> handleEntityNotFoundException(EntityNotFoundException ex) {
+        HttpStatus status = HttpStatus.NOT_FOUND;
+        Map<String, Object> body = createErrorMessageBody(status, ex.getMessage());
+        return new ResponseEntity<>(body, status);
+    }
+
+    @ExceptionHandler(InvalidPriceException.class)
+    protected ResponseEntity<Object> handleInvalidPriceException(InvalidPriceException ex) {
+        HttpStatus status = HttpStatus.BAD_REQUEST;
+        Map<String, Object> body = createErrorMessageBody(status, ex.getMessage());
+        return new ResponseEntity<>(body, status);
+    }
```
duplicate code, we will remove the price ex in that solution, but for the future - if you have the same logic of handling some exceptions just have a method like that
```java
    @ExceptionHandler({
        CustomEx1.class,
        CustomEx2.class})
    public ResponseEntity<Object> handleBadRequestException(RuntimeException ex) {
        HttpStatus status = HttpStatus.BAD_REQUEST;
        var body = createErrorMessageBody(status, ex.getMessage());
        return new ResponseEntity<>(body, status);
    }
```

karma-o commented on src/main/resources/application.properties:
```diff
@@ -4,7 +4,7 @@ spring.datasource.username=${login}
 spring.datasource.driver-class-name=com.mysql.cj.jdbc.Driver
 
-spring.jpa.hibernate.ddl-auto=validate
+spring.jpa.hibernate.ddl-auto=create-drop
```
let's set it back to `validate`

karma-o commented on src/main/resources/db/changelog/changes/01-create-books-table.yaml:
```diff
@@ -29,6 +29,7 @@ databaseChangeLog:
                   type: varchar(255)
                   constraints:
                     nullable: false
+                    unique: true
```
we also do not have any validation for the uniqueness of that `ISBN`, which we could add as a check to the service `create` & `update` methods, because we do not want to return a raw SQL constraint violation exception in case the user tries to save the book with an ISBN that is already present in the DB

karma-o reviewed (changes requested):
Looks good, but we need to fix a couple more things

JohnGrey17 committed:
deleted InvalidPriceException
deleted annotation @Digit for BigInteger variable

JohnGrey17 committed:
deleted InvalidPriceException
deleted annotation @Digit for BigInteger variable
added custom validator for isbn

karma-o commented on src/main/java/org/example/bookstore/validator/UniqueIsbnValidator.java:
```diff
@@ -0,0 +1,22 @@
+import lombok.RequiredArgsConstructor;
+import org.example.bookstore.service.BookService;
+
+@RequiredArgsConstructor
+public class UniqueIsbnValidator implements ConstraintValidator<UniqueIsbn, String> {
+
+    private final BookService bookService;
```
read my comment about that validation once again, I said that we would want this validation in other service methods exactly to avoid that 😅 
Having services/repositories in the DTO validators is a bad practice, usually annotation-based validations are very lightweight and used for some basic things like formats, but anything that would need a db check is done in the service methods like `create`/`update`. Also this won't work because what if you would want to update a book and leave the same ISBN that was before - you'd have to send the same one but it would already exist in the DB, so you'll need to also check if it is different from the one that we already have in the `updateBook()` method

karma-o reviewed (changes requested):
One last thing that we need to fix, feel free to ask questions if you have any, we also covered this topic on some of the last QnA

JohnGrey17 committed:
fixed solution

JohnGrey17 committed:
added addSpecification method

JohnGrey17 committed:
added check in service layer on ISBN uniq

karma-o commented on src/main/java/org/example/bookstore/service/impl/BookServiceImpl.java:
```diff
@@ -22,7 +22,11 @@ public class BookServiceImpl implements BookService {
     @Override
-    public BookDto save(BookRequestDto requestDto) {
+    public BookDto create(BookRequestDto requestDto) {
+        String isbn = requestDto.getIsbn();
+        if (isbn != null && bookRepository.findByIsbn(isbn).isPresent()) {
+            throw new EntityNotFoundException("Book with ISBN " + isbn + " already exists.");
+        }
```
code duplication

karma-o commented on src/main/java/org/example/bookstore/service/impl/BookServiceImpl.java:
```diff
@@ -61,16 +65,16 @@ public BookDto updateBookById(Long id, BookRequestDto updatedBookDto) {
                 .toList();
     }
+
+    @Override
+    public void deleteById(Long id) {
+        bookRepository.deleteById(id);
+    }
```
BTW we should also have an id chek here before preforming the deletion

karma-o reviewed (approved):
GJ!
